blog/content/post/2020-06-30-trenchboot-grub-cbfs.md

```diff
@@ -0,0 +1,575 @@
+    genroms/pxe.rom                0x43cc0    raw             83456 none
+    etc/sercon-port                0x58300    raw                 8 none
+    test-file                      0x58340    raw               584 LZMA (2067 decompressed)
+    (empty)                        0x585c0    null          5928920 none
+    AGESA                          0x5ffdc0   raw            504032 none
+    (empty)                        0x67af00   null           675480 none
+    apu/amdfw                      0x71fdc0   raw            239872 none
+    (empty)                        0x75a700   null           644760 none
+    bootblock                      0x7f7dc0   bootblock       32768 none
+    ```
+
+    > There should be `test-file` section added with `LZMA` description in
+    compressed column.
+
+6. Flash your platform with prepared image.
+
+    Now you have properly prepared firmware, on which further verification can
+    be performed.
+
+#### GRUB package update
+
+1. Update `nixpkgs` repository.
+
+    It is assumed that you have already downloaded custom
+    [3mdeb/nixpkg](https://github.com/3mdeb/nixpkgs) repository to your NixOS.
+    If yes, then go to `nixpkgs/` directory and pull latest changes. Change
+    branch to `grub_cbfs_lzma`.
+
+    ```
+    $ cd ~/nixpkgs/
+    $ git pull
+    $ git checkout grub_cbfs_lzma
+    ```
+
+2. Rebuild NixOS to apply changes.
+
+    ```
+    $ sudo nixos-rebuild switch -I nixpkgs=~/nixpkgs
+    building Nix...
+    building the system configuration...
+    (...)
+    updating GRUB 2 menu...
+    installing the GRUB 2 boot loader on /dev/sda...
+    Installing for i386-pc platform.
+    Installation finished. No error reported.
+    ```
+
+    > There should be `updating GRUB 2 menu...` log after above command. It
+    indicates that grub has been updated.
+
+3. Reboot system and check if you can boot to NixOS via `Secure Launch` entry.
+
+    ```
+    $ reboot
+    ```
+
+If GRUB update procedure went well, we can move on to *GRUB CBFS and LZMA
+support* verification.
+
+#### GRUB CBFS and LZMA support verification
+
+1. Reboot platform and go to GRUB menu.
+
+    ```
+    $ reboot
+    (...)
+
+    *NixOS - Default
+    NixOS - Secure Launch
+    NixOS - All configurations
+
+    Press enter to boot the selected OS, `e' to edit the commands       
+    before booting or `c' for a command-line.                                                                                                                                                 
+    ```
+
+2. Press `c` to enter command-line. Do not boot to NixOS.
+
+    ```
+    Minimal BASH-like line editing is supported. For the first word, TAB   
+    lists possible command completions. Anywhere else TAB lists possible   
+    device or file completions. ESC at any time exits.
+
+    grub>
+    ```
+
+3. Load necessary modules for CBFS and LZMA support.
+
+    ```
+    grub> insmod /boot/grub/i386-pc/lzma.mod
```

Comment:
Please check if `insmod lzma` is enough. These commands would be different for UEFI platform.

Reply:
BIOS environment is enough for verification. As I remember `insmod lzma` wasn't sufficient